### PR TITLE
feat: Protocol support for input streaming parameters on server

### PR DESCRIPTION
**Pull request by SandPod**
Another step to implement https://github.com/serverpod/serverpod/issues/2338.

### Changes:
- Allow inputs streams to be configured and generated for endpoints.
- Prevent streaming connections to be established towards method call endpoints.
- Add a timeout for closing streams when the server is shutting down (preventing deadlocks in the server shutdown sequence).
- Respond with BadRequest and close connection if message targeting an streaming methods return value is received.

### Next up:
- Handle Nullable and Optional input streams. This will require a change in the protocol so that the client can specify what input streams should be opened.
- Allow dynamic streams to be configured (both input and output).
- Add message version to our messages.
- Create a separate message for parameter stream messages and method stream messages.

## Pre-launch Checklist

- [ ] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [ ] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [ ] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [ ] I added new tests to check the change I am making.
- [ ] All existing and new tests are passing.
- [ ] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes

None - this is only touching the new streaming methods.

### CODE REVIEW

File: packages/serverpod_serialization/lib/src/exceptions.dart [hunk 1]
```diff
@@ -20,3 +20,13 @@ abstract class SerializableException implements SerializableModel, Exception {
...
+
+/// This is an exception that is thrown when a stream is closed with an error
+class StreamClosedWithErrorException implements Exception {
+  /// Const constructor to create a new [StreamClosedWithErrorException]
+  const StreamClosedWithErrorException();
+
+  @override
+  String toString() =>
+      'StreamClosedWithErrorException: Stream closed with error';
+}
```
Comment (SandPod) [inline packages/serverpod_serialization/lib/src/exceptions.dart:32]: Move this to Serverpod.

**Review by Isakdl — APPROVED** [review]
Nice work! I have nothing obvious to comment on, a lot of code though. I think you have covered all important cases in the tests.

I have some generic comments, look at them but LGTM!

File: packages/serverpod/lib/src/server/websocket_request_handlers/method_websocket_request_handler.dart [hunk 2]
```diff
@@ -172,95 +212,292 @@ class MethodWebsocketRequestHandler {
...
+      );
+    }
+
+    streamContext.controller.addError(serializableException);
   }
 
   String _buildStreamKey({
     required String endpoint,
     required String method,
+    String? parameter,
```
[the remark this answers is not in the document]
Comment (SandPod) [reply packages/serverpod/lib/src/server/websocket_request_handlers/method_websocket_request_handler.dart:367]: Hmm, I don't really see the benefit. Since we are actually missing the parameter, we might as well ommit it completely from the stream key when it is not present.

Or am I missing something we gain here?

File: packages/serverpod/lib/src/server/websocket_request_handlers/method_websocket_request_handler.dart [hunk 3]
```diff
@@ -276,31 +513,21 @@ class _MethodStreamManager {
...
       return;
     }
 
     // TODO: Support nullable return types.
-    // Becuase encodeWithType doens't support nullable we can't encode null
+    // Because encodeWithType does not support nullable we can't encode null
     // values.
-    if (methodConnector.returnsVoid == false && result != null) {
+    if (methodConnector.returnType != MethodStreamReturnType.voidType &&
+        result != null) {
```
[the remark this answers is not in the document]
Comment (SandPod) [reply packages/serverpod/lib/src/server/websocket_request_handlers/method_websocket_request_handler.dart:530]: Nice, that would probably have been a neater solution! But Klaudio has actually added support for null values so I can probably remove this completely in another PR. Then I can also add tests for `null` return values.

File: packages/serverpod/lib/src/server/websocket_request_handlers/method_websocket_request_handler.dart [hunk 4]
```diff
@@ -102,6 +132,16 @@ class MethodWebsocketRequestHandler {
...
 
+    if (endpointMethodConnector is! MethodStreamConnector) {
+      server.serverpod.logVerbose(
+        'Endpoint method is not a valid stream method: $message',
+      );
+      return OpenMethodStreamResponse.buildMessage(
+        connectionId: message.connectionId,
+        responseType: OpenMethodStreamResponseType.endpointNotFound,
+      );
+    }
```
[the remark this answers is not in the document]
Comment (SandPod) [reply packages/serverpod/lib/src/server/websocket_request_handlers/method_websocket_request_handler.dart:143]: It can only be either a MethodStreamConnector or a MethodConnector (aka MethodCallConnector).  My thinking was that the error message is a bit different that could be helpful for debugging? For example if you remove all streams from a method it will be converted to a MethodCallConnector but you might still have clients trying to connect to it.

File: packages/serverpod/lib/src/server/websocket_request_handlers/method_websocket_request_handler.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (SandPod) [reply packages/serverpod/lib/src/server/websocket_request_handlers/method_websocket_request_handler.dart:530]: Oh wait, I misunderstood your suggestion. We cant just set it to null because then we won't have that field Serialized with type. We now have support for null values following our special serialization. But I will add that in a separate PR since I would like to add tests for that case at the same time.

File: packages/serverpod/lib/src/server/websocket_request_handlers/method_websocket_request_handler.dart [hunk 5]
```diff
@@ -172,95 +212,292 @@ class MethodWebsocketRequestHandler {
...
+        _handleMethodWithStreamReturn(
+          methodConnector: methodConnector,
+          session: session,
+          args: args,
+          streamParams: streamParams,
+          message: message,
+          server: server,
+        );
+        break;
+      case MethodStreamReturnType.singleType:
```
[the remark this answers is not in the document]
Comment (SandPod) [reply packages/serverpod/lib/src/server/websocket_request_handlers/method_websocket_request_handler.dart]: Renamed to `futureType`
